File: listofbool.py
#!/usr/bin/env python3
from time import perf_counter

def ceil(D, d) :
    # returns ceil(D/d), without using floats
    q,r = divmod(D, d)
    if not r :
        return q
    return q+1

bornesup = int(input("Borne supérieure de recherche (exclue) : "))
maxsearch = int(bornesup**0.5)

a = perf_counter()

nombres = [False,] * bornesup # False pour "pas barré"
p = 2

while p < maxsearch :
    nombres[p**2::p] = (True,) * ceil(bornesup-p**2,p)
    p = nombres.index(False, p+1)


print(f"Il y a {nombres.count(False)-2:_} nombres premiers < {bornesup:_}")
b = perf_counter()
# -2 parce qu'on a inclus 0 et 1 dans le comptage
print(f"Temps d'exécution : {b-a:_.2} s")
del nombres

File: listofbool.c
#include <stdlib.h>
#include <stdio.h>

#define usli unsigned long int
#define usc unsigned char


int main() {
  usli bornesup, maxsearch=0, p, totprem=0;
  usc *nombres = NULL;
  printf("Borne supérieure (exclue) : ");
  scanf("%ld", &bornesup);
  do {
    maxsearch ++;
  } while (maxsearch*maxsearch <= bornesup);
  maxsearch --;
  nombres = malloc(bornesup*sizeof(usc));
  if (nombres == NULL) {
    exit(1);
  }
  for (int i=0; i<maxsearch; i++) {
    nombres[i] = 0;
  }

  for (p=2; p<maxsearch; p++) {
    if (nombres[p]) {
      continue;
    }
    for (int i=p*p; i<bornesup; i+=p){
      nombres[i] = 1;
    }
  }
  for (p=2; p<bornesup; p++) {
    if (nombres[p] == 0) {totprem++;}
  }
  free(nombres);
  printf("Il y a %ld nombres premiers inférieurs à %ld\n",
  totprem, bornesup);
}
